mypy/checker.py

```diff
@@ -4053,6 +4095,120 @@ def replay_lookup(new_parent_type: ProperType) -> Optional[Type]:
 
         return output
 
+    def refine_identity_comparison_expression(self,
+                                              operands: List[Expression],
+                                              operand_types: List[Type],
+                                              chain_indices: List[int],
+                                              narrowable_operand_indices: Set[int],
+                                              ) -> Tuple[TypeMap, TypeMap]:
+        """Produces conditional type maps refining expressions used in an identity comparison.
+
+        The 'operands' and 'operand_types' lists should be the full list of operands used
+        in the overall comparison expression. The 'chain_indices' list is the list of indices
+        actually used within this identity comparison chain.
+
+        So if we have the expression:
+
+            a <= b is c is d <= e
+
+        ...then 'operands' and 'operand_types' would be lists of length 5 and 'chain_indices'
+        would be the list [1, 2, 3].
+
+        The 'narrowable_operand_indices' parameter is the set of all indices we are allowed
+        to refine the types of: that is, all operands that will potentially be a part of
+        the output TypeMaps.
+        """
+        singleton = None  # type: Optional[ProperType]
+        possible_singleton_indices = []
+        for i in chain_indices:
+            coerced_type = coerce_to_literal(operand_types[i])
+            if not is_singleton_type(coerced_type):
+                continue
+            if singleton and not is_same_type(singleton, coerced_type):
+                # We have multiple disjoint singleton types. So the 'if' branch
+                # must be unreachable.
+                return None, {}
+            singleton = coerced_type
+            possible_singleton_indices.append(i)
+
+        # There's nothing we can currently infer if none of the operands are singleton types,
+        # so we end early and infer nothing.
+        if singleton is None:
+            return {}, {}
+
+        # If possible, use an unassignable expression as the singleton.
+        # We skip refining the type of the singleton below, so ideally we'd
+        # want to pick an expression we were going to skip anyways.
+        singleton_index = -1
+        for i in possible_singleton_indices:
+            if i not in narrowable_operand_indices:
+                singleton_index = i
+
+        # Oh well, give up and just arbitrarily pick the last item.
+        if singleton_index == -1:
+            singleton_index = possible_singleton_indices[-1]
```

Comment:
Do we have a test for this situation?

Reply:
Currently, no -- but I did spend some time trying to construct one and eventually ended up convincing myself that we'll always get the same result no matter which index we pick. I updated the comment to include the reasoning.